### ISSUE TITLE
[question] Migration to gitlab.freedekstop.org

### ISSUE THREAD

**Issue by bentiss** [libratbag/libratbag#924]
Cc: @stephanlachnit @FFY00 @whot

With @FFY00 we are trying really hard to get regression tests of the various firmwares.
We are currently working on this topic in the [ratbag-emu](https://github.com/libratbag/ratbag-emu) project.

The problem we have, is that for the project to be able to run the tests, we need a VM. We do not want to emulate devices on a bare host, but in a VM. And a VM in a container is even better. This is something we actually already do in [libinput](https://gitlab.freedesktop.org/libinput/libinput).

Which means we already have all of the infrastructure ready to run those tests directly in gitlab.

On the contrary, this is to my knowledge not possible to do on github. There is a self-hosted github actions runner we *could* use, but it is nowhere close to what gitlab runners offer: a clean docker or kubernetes integration to start a new (privileged) container for any job.

We are trying (really hard) in ratbag-emu to keep the code on github and part of the CI in gitlab. This ends up to being very complex, both to design and to maintain. The github API for checks and statuses being marked as a developer preview doesn't help.

Switching ratbag-emu to gitlab would simplify the whole handling of the CI, in a common and not hacked version.

Now, comes the hard question: once we get ratbag-emu in a good shape that it can be valuable for libratbag, do we want to switch libratbag to gitlab too? This would allow to actually run a CI on the drivers, which is something we are terribly lacking right now, as we are introducing more and more regressions.

We still have a little bit of time to take the decision. Note though that this is a technical decision based on an actual need, so this can be easily justified to our users.

PS: yes, of course, we would keep the github part as a mirror.

**Comment by FFY00** [comment]
Does libinput run the CI automatically on PRs? I am trying to access if we should or not do that, the security implications have probably already been considered for libinput.

If we don't run tests automatically on PRs, we will need to have a bot for maintainers to manually aprove a CI run. Right? I think we definitely want the CI to run before merging. If this is the case we will have to maintain a bot either way, unless an out of the box solution already exists. If we do have to maintain a bot regardless if we are on github or gitlab, I don't see much reason to move to gitlab.

If it is fine to run the CI on PRs automatically, moving to gitlab makes perfect sense. Personally I am not that much of a fan of gitlab, but maybe that is because I am just not that used to it.

I see how maintaining the CI could be a pain in the future. Right now I have time, so maintaining it isn't that much of a problem, but that could change.

Pros:
  - The CI is easier to maintain (and is more stable)

Cons:
  - Less users
    - more difficult to reach the public
    - probably fewer contributions
  - Initial migration effort investment

Anyone had something to add to the pros/cons list?

One note: Worst-case scenario regarding the CI maintainability -- we would just end up with an unstable and/or unmaintained CI integration. That doesn't really have an impact on the CI itself, that is another problem, which completely unrelated to github vs gitlab. The CI should still be able to run locally. So, even if the integration doesn't work, the maintainers could still run the CI locally.

We have to balance this with probably losing users/contributors by moving to gitlab.

**Comment by stephanlachnit** [comment]
Generally I don't really mind, if it helps why not. Last time I used freedesktop I wasn't a big fan of it since you need these Access Tokens, but maybe it's possible to disable them for a repository?

> do we want to switch libratbag to gitlab too?

If we do this we might also want to migrate Piper to freedesktop, so we can mention and transfer issues easily.

> PS: yes, of course, we would keep the github part as a mirror.

I think there is even a bot that automatically close opened issues on github and refer to gitlab. But when it comes to issues, I have the feeling more people are on github than on freedesktop.

**Comment by FFY00** [comment]
> I think there is even a bot that automatically close opened issues on github and refer to gitlab. But when it comes to issues, I have the feeling more people are on github than on freedesktop.

Benjamin mentioned on irc that we could still take issues here. What I am afraid of is losing contributors.

**Comment by stephanlachnit** [comment]
> Benjamin mentioned on irc that we could still take issues here. What I am afraid of is losing contributors.

I think that's not a very good idea, since you could easily get twice the same error and less testing of different issues from users with the same hardware. There are bots to move issues between gitlab and github, but I don't like it really. And only taking issues on github while we do MRs on gitlab would be kinda strange, but I guess a solution if it's really necessary.

**Comment by FFY00** [comment]
> This ends up to being very complex, both to design and to maintain.

> Switching ratbag-emu to gitlab would simplify the whole handling of the CI, in a common and not hacked version.

I think there's something to be said here. Even though it is complex, and definitely not optimal, I wouldn't put it as "very complex". We are just issuing an http request, using curl, to gitlab.

> The github API for checks and statuses being marked as a developer preview doesn't help.

For me this is the main issue, the fact that we cannot rely on the API. But hopefully, the final API will be set soon. And even with the API declared stable, it could (and probably will) be changed in the future, so this is +1 thing to maintain.

This is not a dealbreaker, but something to consider. Once API is finalized we should be able to rely on it for at least a few years.

**Comment by bentiss** [comment]
>> This ends up to being very complex, both to design and to maintain.
>>
>> Switching ratbag-emu to gitlab would simplify the whole handling of the CI, in a common and not hacked version.
>
> I think there's something to be said here. Even though it is complex, and definitely not optimal, I wouldn't put it as "very complex". We are just issuing an http request, using curl, to gitlab.

It is "simple" now, but not reliable and flexible. A better approach would be to start the pipeline from github, then monitor each job and provide some reporting within github. This would also allo to use the produced artifacts from gitlab in the github CI, which solves the "upload to codecov" issue much nicer than the current attempts.

> Does libinput run the CI automatically on PRs? I am trying to access if we should do that or not, the security implications have probably already been considered for libinput.

right now the attack target of libinput is rather small. So yes, for every PR, the CI is run.

However, with gitlab we can now have the CI defined in a separate repo with trusted developers. Which means that the CI would run for every PR, but the code run is controlled by a smaller group, not every random submission.

> bot for initiating PR:
yes, maybe. There is marge-bot which seems really nice for what we want. Still it's a setup to handle.

> One note: Worst-case scenario regarding the CI maintainability -- we would just end up with an unstable and/or unmaintained CI integration. That doesn't really have an impact on the CI itself, that is another problem, which completely unrelated to github vs gitlab. The CI should still be able to run locally. So, even if the integration doesn't work, the maintainers could still run the CI locally.

I tend to disagree. If the CI is broken on PR, there is a high chance it won't be run locally by maintainers. See the number of time we fail at running `flake8` until we enforced it in the CI...

>  Last time I used freedesktop.org's gitlab I wasn't a big fan of it since you need these Access Tokens, but maybe it's possible to disable them for a repository?

not sure which project needed that, but a well configured project should not require a specific configuration for forks...

And last (I got to run), IMO, the mirroring issues are something that would need discussed later, once the technical side on whether we need the switch or not.

**Comment by whot** [comment]
> Does libinput run the CI automatically on PRs? I am trying to access if we should do that or not, the security implications have probably already been considered for libinput.

The CI is a first-class citizen in gitlab, so yes, the CI gets run every time you push a branch, the result then gets displayed on the MR page (if any) and then the CI gets re-run after merging to master. A useful thing is that you can get CI results before filing a merge request, saves you a lot of the force-push-and-notify-everyone cycle you need otherwise. There are some conditions you can put into the CI to run things e.g. only on merge to master, or only when $USER is 'foo', etc.

There's a side-effect to gitlab's CI: it's the **only** CI gitlab does by default (you can hook up circle I think but meh). Which means any knowledge translates into other projects on gitlab. This is not always the case with github where we have travis, semphore, circle, etc.

As for the security implications: can you specify what exactly you're worried about here that is different to what we already have here? The containers and VMs are thrown away after a CI run and while you could hook up your CI to do whatever in your own branch, it'd still require a maintainer to pull it into master eventually. And if you wanted to set up a spam bot, I'm sure there are more efficient solutions than a gitlab PR :)

**Comment by bentiss** [comment]
Just to add a little bit of arguments regarding the current custom/curl solution, this is my inbox this morning: 
![Screenshot from 2020-02-14 08-57-01](https://user-images.githubusercontent.com/2705518/74513084-12173700-4f0a-11ea-81cb-4002310bde80.png).

So yes, it "works", but when it breaks, I'd rather not be the one getting the email storm :)

(not blaming anybody here, except myself for not having ironed out the situation)

**Comment by FFY00** [comment]
Yes, sorry. I was playing with travis-ci and totally forgot about that.

It looks like travis had virtualization support. `/proc/cpuinfo` contains `vmx` and other people have had success running virtual machines on it.

I am going to try and setup the ci and see if it works without much trouble.

**Comment by FFY00** [comment]
So, travis supports KVM but it isn't reliable. Unfortunately, I don't think it is a proper alternative :/
